Updates type signature for Field() constructor

bjmc opened the pull request:
Replaces `float` in `lt`/`gt`/`le`/`ge`, arguments with the appropriate Protocol from `annotated_types`.

<!-- Thank you for your contribution! -->
<!-- Unless your change is trivial, please create an issue to discuss the change before creating a PR -->

## Change Summary

Related to #9472 updates the type signatures in `Field()` to be less restrictive, using the appropriate protocols from `annotated_types` instead of strict `float` or `int` types. This should be a nonbreaking change.

## Related issue number

fixes #9472

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

bjmc commented:
The only [test failures](https://github.com/pydantic/pydantic/actions/runs/9195205949/job/25290459935) look to be with an older version of mypy. Maybe it doesn't fully recognize the protocols from `annotated_types`?

sydney-runkle reviewed (approved):
@bjmc,

Thanks, this looks great. Let's consult with @dmontagu re the test failures.

sydney-runkle commented on pydantic/fields.py:
```diff
@@ -56,14 +56,14 @@ class _FromFieldInfoInputs(typing_extensions.TypedDict, total=False):
+    le: annotated_types.SupportsLe | None
+    multiple_of: annotated_types.MultipleOf | None
     strict: bool | None
-    min_length: int | None
-    max_length: int | None
+    min_length: annotated_types.MinLen | None
+    max_length: annotated_types.MaxLen | None
```
For these, let's actually keep the original type (`float` and `int`) and add an additional union member as the `annotated_types` type. Ex:
`multiple_of: int | annotated_types.MultipleOf | None`

sydney-runkle replied:
@bjmc,

Thinking about this more, let's also add some tests accordingly, verifying that `MultipleOf` works in this case (it might not, due to expectations on the `pydantic-core` end of things.

dmontagu commented:
@bjmc the `SupportsX` look like protocols that should work (though as @sydney-runkle mentioned, we should confirm that they actually do work).

But the `MinLen`, `MaxLen`, and `MultipleOf` seem to be dataclasses and not protocols, right? (Am I missing something?) Did you check if instances of these classes actually work? (I'd expect that they won't without further changes, but it's been a while since I looked at these bits of code.)

bjmc commented:
@dmontagu Good shout, I'll revert those.

bjmc commented:
I was overgeneralizing from the `metadata_lookup` [here](https://github.com/pydantic/pydantic/blob/afe67c0a9af244ef2abb726810c5e5db3c680641/pydantic/fields.py#L180) but you're absolutely right those are dataclasses to store the constraints, not Protocols to describe input types. I've reverted them now. 

I suppose to be extremely pedantic, min/max lengths should be typed `Annotated[int, Ge(0)]` rather than just `int` but maybe we don't care about that.

bjmc commented:
@sydney-runkle - Confirming this fixes my issue, using the same `example.py` from #9472 :

```
bjmc@host:~/Sandbox/pydantic$ git checkout main 
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
bjmc@host:~/Sandbox/pydantic$ pyright example.py 
/home/bjmc/Sandbox/pydantic/example.py
  /home/bjmc/Sandbox/pydantic/example.py:6:44 - error: Argument of type "timedelta" cannot be assigned to parameter "gt" of type "float | None" in function "Field"
    Type "timedelta" is incompatible with type "float | None"
      "timedelta" is incompatible with "float"
      "timedelta" is incompatible with "None" (reportArgumentType)
1 error, 0 warnings, 0 informations 
bjmc@host:~/Sandbox/pydantic$ git checkout comparison-protocols 
Switched to branch 'comparison-protocols'
Your branch is up to date with 'origin/comparison-protocols'.
bjmc@host:~/Sandbox/pydantic$ pyright example.py 
0 errors, 0 warnings, 0 informations 
```

sydney-runkle commented:
@bjmc,

Nice, the new changes look good. I think we do still want to add some explicit tests for the new protocols, but I'm not sure what the best way to do that is. I'll consult with @dmontagu and get back to you soon!
